### PR TITLE
C17 Whales - Kiana Shui

### PR DESCRIPTION
**Pull request by kianashui**

### CODE REVIEW

File: swap_meet/clothing.py [hunk 1]
```diff
@@ -1,2 +1,7 @@
-class Clothing:
-    pass
\ No newline at end of file
+from swap_meet.item import Item
+
+class Clothing(Item):
+    def __init__(self, category = "Clothing", condition=0, age=0):
+        super().__init__(category, condition, age)
```
Comment (jericahuang) [inline swap_meet/clothing.py:5]: Great! Taking advantage of inheritance by using the parent class's constructor makes this concise!

File: swap_meet/vendor.py [hunk 2]
```diff
@@ -1,2 +1,87 @@
...
+    def get_by_category(self, category):
+        category_items = []
+        for item in self.inventory:
+            if item.category == category:
+                category_items.append(item)
+
+        return category_items
+    
+    def swap_items(self, other_vendor, my_item, their_item):
+        try:
```
Comment (jericahuang) [inline swap_meet/vendor.py:27]: Awesome use of a try-except!

File: swap_meet/vendor.py [hunk 3]
```diff
@@ -1,2 +1,87 @@
...
+            return False
+
+        return True
+
+    def swap_first_item(self, other_vendor):
+        try:
+            my_item = self.inventory[0]
+            their_item = other_vendor.inventory[0]
+            result = self.swap_items(other_vendor, my_item, their_item)
+            return result
```
Comment (jericahuang) [inline swap_meet/vendor.py:46]: Nice use of .swap_items() return value of True or False!

File: swap_meet/vendor.py [hunk 4]
```diff
@@ -1,2 +1,87 @@
...
+                if item.condition > max_condition:
+                    max_condition = item.condition
+                    best_condition = item
+        
+        return best_condition
+
+    def swap_best_by_category(self, other, my_priority, their_priority):
+        my_best = self.get_best_by_category(their_priority)
+        their_best = other.get_best_by_category(my_priority)
+        if my_best == None or their_best == None:
```
Comment (jericahuang) [inline swap_meet/vendor.py:64]: When checking for None values, it's better convention to use `is` rather than `==`

File: swap_meet/vendor.py [hunk 5]
```diff
@@ -1,2 +1,87 @@
...
+            return False
+        else:
+            self.swap_items(other, my_best, their_best)
+            return True
+    
+    def get_newest_item(self):
+        max_age = 0
```
Comment (jericahuang) [inline swap_meet/vendor.py:71]: Nice implementation of this optional enhancement! Some people might interpret newest to be the minimum age, but it's the logic and implementation that matters, and you did great!

File: swap_meet/vendor.py [hunk 6]
```diff
@@ -1,2 +1,87 @@
...
+            if item.age > max_age:
+                max_age = item.age
+                newest_item = item
+        return newest_item
+
+    def swap_by_newest(self, other_vendor):
+        my_newest_item = self.get_newest_item()
+        their_newest_item = other_vendor.get_newest_item()
+
+        if my_newest_item is None or their_newest_item is None:
```
Comment (jericahuang) [inline swap_meet/vendor.py:83]: 👍🏻

File: tests/unit_tests/test_wave_01.py [hunk 7]
```diff
@@ -40,16 +36,12 @@ def test_removing_from_inventory_returns_item():
...
     )
 
     result = vendor.remove(item)
-
-    raise Exception("Complete this test according to comments below.")
-    # *********************************************************************
-    # ****** Complete Assert Portion of this test **********
-    # *********************************************************************
+    
+    assert result == False
```
Comment (jericahuang) [inline tests/unit_tests/test_wave_01.py:47]: 👍🏻

File: tests/unit_tests/test_wave_02.py [hunk 8]
```diff
@@ -34,7 +31,7 @@ def test_get_no_matching_items_by_category():
...
     items = vendor.get_by_category("electronics")
 
-    raise Exception("Complete this test according to comments below.")
-    # *********************************************************************
-    # ****** Complete Assert Portion of this test **********
-    # *********************************************************************
+    assert len(items) == 0
+    assert item_a not in items
+    assert item_b not in items
+    assert item_c not in items
```
Comment (jericahuang) [inline tests/unit_tests/test_wave_02.py:37]: 💯

File: tests/unit_tests/test_wave_06.py [hunk 9]
```diff
@@ -76,16 +72,17 @@ def test_swap_best_by_category():
...
-    # - That all the correct items are in tai and jesse's inventories, including the items which were swapped from one vendor to the other
+    # Assert
+    assert result == True
+    assert len(tai.inventory) == 3
+    assert len(jesse.inventory) == 3
+    assert item_a in tai.inventory
+    assert item_b in tai.inventory
+    assert item_f in tai.inventory
+    assert item_d in jesse.inventory
+    assert item_e in jesse.inventory
```
Comment (jericahuang) [inline tests/unit_tests/test_wave_06.py:83]: 👍🏻

File: tests/unit_tests/test_wave_06.py [hunk 10]
```diff
@@ -227,11 +224,189 @@ def test_swap_best_by_category_no_other_match_is_false():
...
-    raise Exception("Complete this test according to comments below.")
-    # *********************************************************************
-    # ****** Complete Assert Portion of this test **********
-    # *********************************************************************
-    # Assertions should check:
-    # - That result is falsy
-    # - That tai and jesse's inventories are the correct length
-    # - That all the correct items are in tai and jesse's inventories
+    # Assert
+    assert result == False
```
Comment (jericahuang) [inline tests/unit_tests/test_wave_06.py:228]: This works! Falsy values are values that evaluate to False in a boolean context, such as empty sequences (lists, tuples, strings, dictionaries, sets), zero, and None, in addition to False. So to check if something is falsy, `assert not result`

File: tests/unit_tests/test_wave_06.py [hunk 11]
```diff
@@ -227,11 +224,189 @@ def test_swap_best_by_category_no_other_match_is_false():
...
+    item_c = Decor(condition=4.0)
+    tai = Vendor(
+        inventory=[item_a, item_b, item_c]
+    )
+
+    newest_item = tai.get_newest_item()
+
+    assert newest_item is None
+
+def test_newest_item_with_duplicates():
```
Comment (jericahuang) [inline tests/unit_tests/test_wave_06.py:272]: Great test to check for behavior when there are duplicates!

File: tests/unit_tests/test_wave_06.py [hunk 12]
```diff
@@ -227,11 +224,189 @@ def test_swap_best_by_category_no_other_match_is_false():
...
+    assert item_d in jesse.inventory
+    assert item_e in jesse.inventory
+    assert item_f in jesse.inventory
+
+
+
+
+
+
+####optional:
```
Comment (jericahuang) [inline tests/unit_tests/test_wave_06.py:243]: Very great thorough tests for the optional enhancement!

File: tests/unit_tests/test_wave_06.py [hunk 13]
```diff
@@ -227,11 +224,189 @@ def test_swap_best_by_category_no_other_match_is_false():
...
+    assert len(tai.inventory) == 3
+    assert len(jesse.inventory) == 3
+    assert item_a in tai.inventory
+    assert item_d in tai.inventory
+    assert item_c in tai.inventory
+    assert item_b in jesse.inventory
+    assert item_e in jesse.inventory
+    assert item_f in jesse.inventory
+
+def test_swap_by_newest_reordered():
```
Comment (jericahuang) [inline tests/unit_tests/test_wave_06.py:321]: 💯

File: tests/unit_tests/test_wave_06.py [hunk 14]
```diff
@@ -227,11 +224,189 @@ def test_swap_best_by_category_no_other_match_is_false():
...
+    result = tai.swap_by_newest(jesse)
+
+    assert not result
+    assert len(tai.inventory) == 0
+    assert len(jesse.inventory) == 3
+    assert item_a in jesse.inventory
+    assert item_b in jesse.inventory
+    assert item_c in jesse.inventory
+
+def test_swap_by_newest_no_other_inventory_is_false():
```
Comment (jericahuang) [inline tests/unit_tests/test_wave_06.py:375]: Awesome edge case tests!

**Review by jericahuang — COMMENTED** [review]
Phenomenal work on this project, Kiana! You demonstrated mastery of Unit 1 concepts and did fantastic with the optional enhancements. 👍🏻 Your project is green 🟢 !
